File: 038. Count and Say/Count and Say.java
public class Solution {
    public String countAndSay(int n) {
        if(n == 1){
            return "1";
        }
        //递归调用，然后对字符串处理
        String str = countAndSay(n-1) + "*";//为了str末尾的标记，方便循环读数
        char[] c = str.toCharArray();
        int count = 1;
        String s = "";
        for(int i = 0; i < c.length - 1;i++){
        	if(c[i] == c[i+1]){
        		count++;//计数增加
        	}else{
        		s = s + count + c[i];//上面的*标记这里方便统一处理
        		count = 1;//初始化
        	}
        }
        return s;
    }
}

！

File: 038. Count and Say/count and say.py
'''
用一个下标来表示当前统计的字符的起始位置，一个计数器来表示该字符的数目。
不断读取直到字符不相等，添加到结果集中，更新起始位置和计数器。下面代码中的计数器用下标相减代替。
'''

class Solution(object):
    def countAndSay(self, n):
        """
        :type n: int
        :rtype: str
        """
        result = "1"
        for __ in range(1, n):
            result = self.getNext(result)
        return result

    def getNext(self, s):
        result = []
        start = 0
        while start < len(s):
            curr = start + 1
            while curr < len(s) and s[start] == s[curr]:
                curr += 1
            result.extend((str(curr - start), s[start]))
            start = curr
        return "".join(result)


if __name__ == "__main__":
    assert Solution().countAndSay(4) == "1211"
    assert Solution().countAndSay(5) == "111221"
